src/hotspot/share/logging/logTagSet.cpp

```diff
@@ -124,16 +124,31 @@ void LogTagSet::vwrite(LogLevelType level, const char* fmt, va_list args) {
     ret = os::vsnprintf(buf, sizeof(buf), fmt, args);
   }
   assert(ret >= 0, "Log message buffer issue");
-  if ((size_t)ret >= sizeof(buf)) {
-    size_t newbuf_len = prefix_len + ret + 1;
-    char* newbuf = NEW_C_HEAP_ARRAY(char, newbuf_len, mtLogging);
-    prefix_len = _write_prefix(newbuf, newbuf_len);
-    ret = os::vsnprintf(newbuf + prefix_len, newbuf_len - prefix_len, fmt, saved_args);
-    assert(ret >= 0, "Log message buffer issue");
-    log(level, newbuf);
-    FREE_C_HEAP_ARRAY(char, newbuf);
-  } else {
+  size_t newbuf_len = (size_t)ret + prefix_len + 1; // total bytes needed including prefix.
+  if (newbuf_len <= sizeof(buf)) {
     log(level, buf);
+  } else {
+    // Buffer too small, allocate a large enough buffer by using of malloc/free to avoid circularity.
+    char* newbuf = (char*)::calloc(newbuf_len, sizeof(char));
+    if (newbuf != nullptr) {
+      prefix_len = _write_prefix(newbuf, newbuf_len);
+      ret = os::vsnprintf(newbuf + prefix_len, newbuf_len - prefix_len, fmt, saved_args);
+      assert(ret >= 0, "Log message buffer issue");
+      log(level, newbuf);
+      ::free(newbuf);
+    } else {
```

Comment:
If we failed to allocate newbuf the existing buf is already filled with everything we need. Can't we just overwrite the last part with the "truncated" message rather than re-doing the whole thing?

Reply:
line 124:  When buf is small, buf is filled with whatever truncated string with format to get the length needed for new buffer, it does not include prefix. The new version keeps prefix and the buf ends with "..(truncated), native OOM" even if prefix already overflows buf.

Reply:
Ah I missed that.  But do we need to do that? If we introduce:
`char buf2[1];`
at line 123 and use that to determine the necessary additional length, then we don't need to re-start the whole process from scratch.

Reply:
I agree, that would be easier:
- at line 124, print into a dummy buffer of len 1. That gives you the length of the expanded string, but leaves the (truncated) prefix in buf intact. According to C99 a one should even be able to use a `NULL, 0` for the buffer but I am not sure where we are with os::vnsprintf.
- Then, at line 140ff, buf contains either just the truncated prefix (line 124) or the truncated prefix + message (122). You then just stamp your truncated message over it. 

Only, make sure (maybe static assert) that the truncation text is smaller than the buffer :)

Reply:
Thanks, it is using ::vnsprintf in os::vnsprintf, tested on linux, using NULL is OK but Windows counterpart is quite complex --- should avoid using 0 as length. I will be careful to use args and saved_args, they should only be used once --- it could be altered by call to ::vnsprintf. The patch already used it twice, now with David's suggestion, that usage will be eliminated.